File: guides/micronaut-object-storage-aws/java/src/main/java/example/micronaut/ProfilePicturesController.java
package example.micronaut;

import io.micronaut.http.HttpHeaders;
import io.micronaut.http.HttpRequest;
import io.micronaut.http.HttpResponse;
import io.micronaut.http.MediaType;
import io.micronaut.http.MutableHttpResponse;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Delete;
import io.micronaut.http.annotation.Get;
import io.micronaut.http.annotation.Post;
import io.micronaut.http.multipart.CompletedFileUpload;
import io.micronaut.http.server.types.files.StreamedFile;
import io.micronaut.http.server.util.HttpHostResolver;
import io.micronaut.http.uri.UriBuilder;
import io.micronaut.objectstorage.aws.AwsS3ObjectStorageEntry;
import io.micronaut.objectstorage.aws.AwsS3Operations;
import io.micronaut.objectstorage.request.UploadRequest;
import io.micronaut.objectstorage.response.UploadResponse;
import io.micronaut.scheduling.TaskExecutors;
import io.micronaut.scheduling.annotation.ExecuteOn;
import software.amazon.awssdk.services.s3.model.GetObjectResponse;
import software.amazon.awssdk.services.s3.model.ObjectCannedACL;
import software.amazon.awssdk.services.s3.model.PutObjectResponse;

import java.net.URI;
import java.util.Optional;

//tag::begin-class[]
@Controller(ProfilePicturesController.PREFIX) // <1>
@ExecuteOn(TaskExecutors.IO) // <2>
public class ProfilePicturesController implements ProfilePicturesApi {

    static final String PREFIX = "/pictures";

    private final AwsS3Operations objectStorage; // <3>
    private final HttpHostResolver httpHostResolver; // <4>

    public ProfilePicturesController(AwsS3Operations objectStorage, HttpHostResolver httpHostResolver) {
        this.objectStorage = objectStorage;
        this.httpHostResolver = httpHostResolver;
    }
//end::begin-class[]

    //tag::upload[]
    @Override
    public HttpResponse<?> upload(CompletedFileUpload fileUpload, String userId, HttpRequest<?> request) {
        String key = buildKey(userId); // <1>
        UploadRequest objectStorageUpload = UploadRequest.fromCompletedFileUpload(fileUpload, key); // <2>
        UploadResponse<PutObjectResponse> response = objectStorage.upload(objectStorageUpload, builder -> {  // <3>
            builder.acl(ObjectCannedACL.PUBLIC_READ); // <4>
        });

        return HttpResponse
                .created(location(request, userId)) // <5>
                .header(HttpHeaders.ETAG, response.getETag()); // <6>
    }

    private static String buildKey(String userId) {
        return userId + ".jpg";
    }

    private URI location(HttpRequest<?> request, String userId) {
        return UriBuilder.of(httpHostResolver.resolve(request))
                .path(PREFIX)
                .path(userId)
                .build();
    }
    //end::upload[]

    //tag::download[]
    @Override
    public Optional<HttpResponse<StreamedFile>> download(String userId) {
        String key = buildKey(userId);
        return objectStorage.retrieve(key) // <1>
                .map(ProfilePicturesController::buildStreamedFile); // <2>
    }

    private static HttpResponse<StreamedFile> buildStreamedFile(AwsS3ObjectStorageEntry entry) {
        GetObjectResponse nativeEntry = entry.getNativeEntry();
        MediaType mediaType = MediaType.of(nativeEntry.contentType());
        StreamedFile file = new StreamedFile(entry.getInputStream(), mediaType).attach(entry.getKey());
        MutableHttpResponse<Object> httpResponse = HttpResponse.ok()
                .header(HttpHeaders.ETAG, nativeEntry.eTag()); // <3>
        file.process(httpResponse);
        return httpResponse.body(file);
    }
    //end::download[]

    //tag::delete[]
    @Override
    public void delete(String userId) {
        String key = buildKey(userId);
        objectStorage.delete(key);
    }
    //end::delete[]

//tag::end-class[]
}
//end::end-class[]

File: guides/micronaut-object-storage-aws/python/src/example/micronaut/profile_pictures_controller.py
from micronaut.http import HttpHeaders, HttpRequest, HttpResponse, HttpStatus, MediaType
from micronaut.http.annotation import Controller, Delete, Get, Post, Status
from micronaut.http.multipart import CompletedFileUpload
from micronaut.http.server.types.files import StreamedFile
from micronaut.http.server.util import HttpHostResolver
from micronaut.http.uri import UriBuilder
from micronaut.objectstorage.aws import AwsS3ObjectStorageEntry, AwsS3Operations
from micronaut.objectstorage.request import UploadRequest
from micronaut.scheduling import TaskExecutors
from micronaut.scheduling.annotation import ExecuteOn
from software.amazon.awssdk.services.s3.model import ObjectCannedACL


# tag::begin-class[]
@Controller("/pictures")  # <1>
@ExecuteOn(TaskExecutors.BLOCKING)  # <2>
class ProfilePicturesController:

    PREFIX = "/pictures"

    def __init__(
        self,
        object_storage: AwsS3Operations,  # <3>
        http_host_resolver: HttpHostResolver,  # <4>
    ):
        self.object_storage = object_storage
        self.http_host_resolver = http_host_resolver
# end::begin-class[]

    # tag::upload[]
    @Post(uri="/{userId}", consumes=MediaType.MULTIPART_FORM_DATA)
    def upload(
        self,
        fileUpload: CompletedFileUpload,
        userId: str,
        request: HttpRequest,
    ) -> HttpResponse:
        key = self.build_key(userId)  # <1>
        object_storage_upload = UploadRequest.fromCompletedFileUpload(fileUpload, key)  # <2>
        response = self.object_storage.upload(  # <3>
            object_storage_upload,
            lambda builder: builder.acl(ObjectCannedACL.PUBLIC_READ),  # <4>
        )

        return (
            HttpResponse.created(self.location(request, userId))  # <5>
            .header(HttpHeaders.ETAG, response.getETag())  # <6>
        )

    @staticmethod
    def build_key(userId: str) -> str:
        return f"{userId}.jpg"

    def location(self, request: HttpRequest, userId: str):
        return (
            UriBuilder.of(self.http_host_resolver.resolve(request))
            .path(self.PREFIX)
            .path(userId)
            .build()
        )
    # end::upload[]

    # tag::download[]
    @Get("/{userId}")
    def download(self, userId: str) -> HttpResponse[StreamedFile]:
        key = self.build_key(userId)
        entry = self.object_storage.retrieve(key)  # <1>
        if entry.isEmpty():
            return HttpResponse.notFound()
        return self.build_streamed_file(entry.get())  # <2>

    @staticmethod
    def build_streamed_file(entry: AwsS3ObjectStorageEntry) -> HttpResponse[StreamedFile]:
        native_entry = entry.getNativeEntry()
        media_type = MediaType.of(native_entry.contentType())
        file = StreamedFile(entry.getInputStream(), media_type).attach(entry.getKey())
        http_response = HttpResponse.ok().header(HttpHeaders.ETAG, native_entry.eTag())  # <3>
        file.process(http_response)
        return http_response.body(file)
    # end::download[]

    # tag::delete[]
    @Status(HttpStatus.NO_CONTENT)
    @Delete("/{userId}")
    def delete(self, userId: str) -> None:
        key = self.build_key(userId)
        self.object_storage.delete(key)
    # end::delete[]

# tag::end-class[]
# end::end-class[]
